### PR TITLE
Backgrounding CSVs 

### PR DESCRIPTION
**Pull request by irosenb**
This is a work in progress. Not ready to merge.

### CODE REVIEW

**Comment by irosenb** [comment]
@dinshaw Am I heading in the right direction?

File: app/models/export.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Export < ActiveRecord::Base
+  has_attached_file :file 
+  before_create :attach_file 
+
+  def attach_file 
+    users = User.all
+    csv = UsersExport(users).to_csv
+    # file = File.
+  end
```
Comment (dinshaw) [inline app/models/export.rb]: If this method is going to kick-off the csv generation, that I would name it as such.
Calling it `attach_file` seems opaque/overloaded

**Comment by dinshaw** [comment]
In general looking very good/clean.
Be very careful with naming, as it will lead you to good design.
Keep in mind the 'single responsibility' idea.
When you get a little further,  lets do a live walk-through/code-review.

**Comment by irosenb** [comment]
@dinshaw should Phran be able to name the Exports?

**Comment by irosenb** [comment]
Also I'm kinda getting the hang of this.

**Comment by dinshaw** [comment]
@irosenb no naming, just a nice, readable timestamp name

**Comment by dinshaw** [comment]
... and glad to hear it! Ready for a review tomorrow?

**Comment by irosenb** [comment]
Yes let's do a review. How about at 5:30 PM ET?

**Comment by dinshaw** [comment]
sounds good.

**Comment by irosenb** [comment]
Skype? @dinshaw

**Comment by irosenb** [comment]
@dinshaw Pretty close to done. Your thoughts?

File: app/controllers/exports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_export
+      @export = Export.find(params[:id])
+    end
+
+    # Only allow a trusted parameter "white list" through.
+    def export_params
+      params[:export]
```
Comment (dinshaw) [inline app/controllers/exports_controller.rb]: What about the strong parameter syntax built into rails for this

File: spec/controllers/exports_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  # This should return the minimal set of values that should be in the session
+  # in order to pass any filters (e.g. authentication) defined in
+  # ExportsController. Be sure to keep this updated too.
+  let(:valid_session) { {} }
+
+  describe "GET index" do
+    it "assigns all exports as @exports" do
+      puts Export.count
```
Comment (dinshaw) [inline spec/controllers/exports_controller_spec.rb]: remove logging

File: spec/controllers/exports_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      export = Export.create! valid_attributes
```
Comment (dinshaw) [inline spec/controllers/exports_controller_spec.rb]: Are you not using the `let(:foo)` syntax on purpose?

File: spec/helpers/exports_helper_spec.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (dinshaw) [inline spec/helpers/exports_helper_spec.rb]: dead file

File: spec/requests/exports_spec.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
```
Comment (dinshaw) [inline spec/requests/exports_spec.rb]: dead file

File: spec/routing/exports_routing_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+require "rails_helper"
```
Comment (dinshaw) [inline spec/routing/exports_routing_spec.rb]: Glad you're writing tests. Feel free to do the routing specs if you find them helpful, but i don't really do them

File: spec/views/exports/index.html.erb_spec.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+RSpec.describe "exports/index", :type => :view do
+  before(:each) do
+    assign(:exports, [
+      Export.create!(),
+      Export.create!()
+    ])
+  end
+
+  it "renders a list of exports" do
+    render
```
Comment (dinshaw) [inline spec/views/exports/index.html.erb_spec.rb]: ?

File: spec/views/exports/new.html.erb_spec.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+RSpec.describe "exports/new", :type => :view do
+  before(:each) do
+    assign(:export, Export.new())
+  end
+
+  it "renders new export form" do
+    render
+
+    assert_select "form[action=?][method=?]", exports_path, "post" do
```
Comment (dinshaw) [inline spec/views/exports/new.html.erb_spec.rb]: ?

File: spec/views/exports/show.html.erb_spec.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe "exports/show", :type => :view do
+  before(:each) do
+    @export = assign(:export, Export.create!())
+  end
+
+  it "renders attributes in <p>" do
+    render
```
Comment (dinshaw) [inline spec/views/exports/show.html.erb_spec.rb]: ?

File: app/controllers/exports_controller.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    if @export
+      redirect_to @export, notice: 'Export was successfully created.'
+    else
+      render :new
+    end
+  end
+
+  # PATCH/PUT /exports/1
+  def update
```
Comment (dinshaw) [inline app/controllers/exports_controller.rb]: What would you do with a update?

File: app/helpers/exports_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module ExportsHelper
```
Comment (dinshaw) [inline app/helpers/exports_helper.rb]: dead file

File: app/admin/users.rb [hunk 13]
```diff
@@ -5,6 +5,11 @@ def scoped_collection
       end_of_association_chain.includes(:business)
     end
   end
+
+  collection_action :export_csv, method: :get do
```
Comment (dinshaw) [inline app/admin/users.rb]: What is this doing?

File: app/controllers/exports_controller.rb (hunk 2 above)
Comment (irosenb) [inline app/controllers/exports_controller.rb]: I don't really have any parameters per se.

File: spec/controllers/exports_controller_spec.rb (hunk 4 above)
Comment (irosenb) [inline spec/controllers/exports_controller_spec.rb]: @dinshaw It was auto generated by `rails g controller`

File: spec/routing/exports_routing_spec.rb (hunk 7 above)
Comment (irosenb) [inline spec/routing/exports_routing_spec.rb]: Yeah. This was auto generated when I ran `rails generate controller`

File: app/controllers/exports_controller.rb (hunk 2 above)
Comment (dinshaw) [inline app/controllers/exports_controller.rb]: `params.require(:export).permit(:foo)`

File: spec/controllers/exports_controller_spec.rb (hunk 4 above)
Comment (dinshaw) [inline spec/controllers/exports_controller_spec.rb]: Please pull out all the useless files. The Rails generators can be helpful, but don't just commit all of them. Only keep what is being usied.

**Comment by dinshaw** [comment]
See my PR against your branch.

**Comment by irosenb** [comment]
@dinshaw should I get rid of the `ExportsController`? Active Admin does the same thing, and having those routes exposed to the public is not a good thing

**Comment by dinshaw** [comment]
Get rid of everything that is not essential (always)

**Comment by irosenb** [comment]
:+1:

**Comment by irosenb** [comment]
:fork_and_knife: ready

**Comment by irosenb** [comment]
@dinshaw should be ready

**Comment by irosenb** [comment]
@dinshaw ready for merge :smile:

**Comment by dinshaw** [comment]
So the code looks ok, went to try it. Spun up the server, and I don't get how it works.
Please add a doc describing how to generate an export, and retrieve it.

**Comment by irosenb** [comment]
Will do. In the meantime:

- run`rake jobs:run`
- go to the admin console
- click on exports
- create a new export
- submit the form
- when the CSV export is done, the export show page will show the file
name.

**Comment by dinshaw** [comment]
Should this task be available to me locally?
```
2.1.2 forever-family-foundation(csv*)$ be rake jobs:run
rake aborted!
Don't know how to build task 'jobs:run'

(See full trace by running task with --trace)
```

**Comment by irosenb** [comment]
Whoops it's rake jobs:work

**Comment by dinshaw** [comment]
seems like it created the export, but i'm getting errors in the DJ log: 
```
2.1.2 forever-family-foundation(csv*)$ be rake jobs:work
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] Starting job worker
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] Job Export#save_csv_without_delay (id=1) RUNNING
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] Job Export#save_csv_without_delay (id=1) FAILED (0 prior attempts) with Net::SMTPAuthenticationError: 530-5.5.1 Authentication Required. Learn more at
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] 1 jobs processed at 0.0423 j/s, 1 failed
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] Job Export#save_csv_without_delay (id=1) RUNNING
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] Job Export#save_csv_without_delay (id=1) FAILED (1 prior attempts) with Net::SMTPAuthenticationError: 530-5.5.1 Authentication Required. Learn more at
[Worker(host:Dinshaws-MacBook-Pro.local pid:44961)] 1 jobs processed at 0.0400 j/s, 1 failed

```

**Comment by irosenb** [comment]
Oh I think you need to load the `env` variables. Try throwing in `gem
'figaro'` into the gemfile.

**Comment by dinshaw** [comment]
Please push complete code. I can't look right now, but I recall you added that gem already

Sent via thousands of miles of curved space-time. 

On Mar 13, 2015, at 09:39, Isaac Rosenberg <notifications@github.com> wrote:

Oh I think you need to load the `env` variables. Try throwing in `gem
'figaro'` into the gemfile.

**Comment by irosenb** [comment]
Ahh I think it might be that you don't have the right `ENV` variables. Do you have the gmail username and password in `application.yml`?

**Comment by dinshaw** [comment]
Nope. If I should have a local config file, commit an example file with dummy data and a doc

Sent via thousands of miles of curved space-time. 

On Mar 13, 2015, at 11:27, Isaac Rosenberg <notifications@github.com> wrote:

Ahh I think it might be that you don't have the right ENV variables. Do you have the gmail username and password in application.yml?

**Comment by irosenb** [comment]
I emailed you a config file. Download that and move it to
`config/application.yml`.

Feels like too much to commit a file.

**Comment by dinshaw** [comment]
Commit it, please. The example file needs to be committed so they people know the structure. Then they can copy it, and update their values. 

Sent via thousands of miles of curved space-time. 

On Mar 14, 2015, at 14:20, Isaac Rosenberg <notifications@github.com> wrote:

I emailed you a config file. Download that and move it to
`config/application.yml`.

Feels like too much to commit a file.

**Comment by irosenb** [comment]
Done

**Comment by dinshaw** [comment]
Thanks. Flying back tonight. Wil try and verify and merge on plane. 

Sent via thousands of miles of curved space-time. 

On Mar 14, 2015, at 15:31, Isaac Rosenberg <notifications@github.com> wrote:

Done

**Comment by dinshaw** [comment]
Testing now, but the convention is to commit example files with their usual extension + .example.
Please make that change.

**Comment by dinshaw** [comment]
did you commit all our production values to the public repo?
If so, they all need to be changed, asap.

**Comment by dinshaw** [comment]
But functionality and implementation look good. 
I'll merge as soon as the above is addressed.

**Comment by irosenb** [comment]
Okay definitely my fault on exposing the keys. I rebased & redid the file.

**Comment by irosenb** [comment]
@dinshaw ready to merge?

**Comment by dinshaw** [comment]
So the config vars in the example file are not our prod creds?

**Comment by irosenb** [comment]
Right now, no. But I had accidentally committed them. Do you want me to
change them?

**Comment by dinshaw** [comment]
Change them? I want you to confirm that there are no prod creds in the repo.

**Comment by irosenb** [comment]
Yes, I can confirm that there are no production creds anymore.

**Pull request merged**